### PR TITLE
Swagger ui redesign/standardize default buttons

### PR DESCRIPTION
**Pull request by Rody-Kirwan**
### Description
A lot of file changes here but not a whole lot going on. I've created and applied some default css button classes to follow the UX spec:
 - sui-btn (default class)
 - sui-btn--sm
 - sui-btn--primary
 - sui-btn--secondary
 - sui-btn--tertiary-lt (ghost button on light bg)
 - sui-btn--tertiary-dk (ghost button on dark bg) 
 - sui-btn-transparent (a tight transparent button wrapper for clickable icons)

 - sui-btn-group 
 - sui-btn-wrapper

Updated any native html `<button>` to use custom `<Button>` component in `layout-utils` 

Added `<span>` wrapper to button-text to maintain `em` margin / padding etc when setting font-size.

### Motivation and Context
This completes the necessary changes for buttons as part of the UX redesign

### How Has This Been Tested?
Manually and existing Cypress test

### Screenshots (if appropriate):
Screen Shot 2019-05-20 at 12 26 27

Screen Shot 2019-05-20 at 12 26 51

Screen Shot 2019-05-20 at 12 28 17

Screen Shot 2019-05-20 at 12 27 53

Screen Shot 2019-05-20 at 12 27 12

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Rody-Kirwan** [comment]
Ping @lshaw-sb @taraokelly

File: src/core/components/auth/authorize-btn.jsx [hunk 1]
```diff
@@ -17,12 +19,18 @@ export default class AuthorizeBtn extends React.Component {
 
     return (
       <div className="auth-wrapper">
-        <button className={isAuthorized ? "btn authorize locked" : "btn authorize unlocked"} onClick={onClick}>
+        <Button
+          className={`sui-btn sui-btn--secondary authorize ${isAuthorized ? "locked" : "unlocked"}`}
```
Comment (Rody-Kirwan) [inline src/core/components/auth/authorize-btn.jsx]: There's a bit of repetition with `sui-btn sui-btn--[moderator]` on _most_ buttons - which I'd prefer to abstract. I was thinking of adding them as default classNames within the `Button` class, with another prop for the moderator - for example: ` mod="primary"`.

Some concerns with this approach is - it limits the `Button` component to these default styles, and also (maybe more importantly) creates a non-familiar interface for a common element. Even so - I'm adding a suggested implementation below - let me know what you think @shockey 

```
// Default Button
<Button
   className={`authorize ${isAuthorized ? "locked" : "unlocked"}`}
   mod="primary"
   onClick={onClick}
 />

// Custom Button
<Button
   className={`authorize ${isAuthorized ? "locked" : "unlocked"}`}
   default=false
   onClick={onClick}
 />
export class Button extends React.Component {

  static propTypes = {
    className: PropTypes.string,
    mod: PropTypes.string,
    default: PropTypes.bool
  }

  static defaultProps = {
    className: "",
    mod: "primary",
    default: true,
  }

  applyDefaultClass = () => this.props.default ? `sui-btn sui-btn--${this.props.mod}` : ""

  render() {
    return (
      <button
        {...this.props}
        className={xclass(this.props.className, "button", this.applyDefaultClass())}
      />
    )
  }
}
```

File: src/core/components/param-body.jsx [hunk 2]
```diff
@@ -123,12 +123,16 @@ export default class ParamBody extends PureComponent {
...
-                         </Button>
-                         </div>
+            !isExecute
+              ? null
+              : <div className="body-param-edit">
+                  <Button
+                    className={`sui-btn sui-btn--tertiary-lt sui-btn--sm body-param__example-edit ${isEditBox ? "disregard" : "edit"}`}
+                    onClick={this.toggleIsEditBox}
+                  >
+                    { isEditBox ? "Disregard" : "Edit"}
```
Comment (Rody-Kirwan) [inline src/core/components/param-body.jsx]: text-change as per UX

File: src/core/components/auth/authorize-btn.jsx (hunk 1 above)
Comment (shockey) [reply src/core/components/auth/authorize-btn.jsx]: **Change request**: let's go with your implementation @Rody-Kirwan, but invert the `default` prop: instead, let's have an optional `unstyled` prop that, if set, omits the default classes.

File: src/core/components/param-body.jsx (hunk 2 above)
Comment (shockey) [reply src/core/components/param-body.jsx]: No problem, not a big fan but will bring it up outside this PR. Approved for the purposes of this PR!

File: src/core/components/operation-tag.jsx [hunk 3]
```diff
@@ -56,8 +57,7 @@ export default class OperationTag extends React.Component {
     let showTag = layoutSelectors.isShown(isShownKey, docExpansion === "full" || docExpansion === "list")
 
     return (
-      <div className={showTag ? "opblock-tag-section is-open" : "opblock-tag-section"} >
-
+      <div className={`opblock-tag-section ${showTag ? "is-open" : ""}`}>
```
Comment (shockey) [inline src/core/components/operation-tag.jsx]: **Change request**: use `classnames` instead of ternary operators here, example: https://github.com/swagger-api/swagger-ui/blob/ad786b023f8e54fcdb4a883039425148d8a8a327/src/core/json-schema-components.jsx#L258

it does have a lot of other usage patterns, see [here](https://github.com/JedWatson/classnames#usage) for documentation!

File: src/core/components/auth/authorize-btn.jsx (hunk 1 above)
Comment (shockey) [reply src/core/components/auth/authorize-btn.jsx]: **Change request**: rename the `authorize` class so that it makes sense w/r/t BEM in the context of this button (perhaps make it a modifier of `sui-btn`? no idea what it's trying to convey)

File: src/core/components/auth/auths.jsx [hunk 4]
```diff
@@ -80,12 +80,13 @@ export default class Auths extends React.Component {
                   />
               }).toArray()
             }
-            <div className="auth-btn-wrapper">
+            <div className="auth-btn-wrapper sui-btn-wrapper">
               {
-                nonOauthDefinitions.size === authorizedAuth.size ? <Button className="btn modal-btn auth" onClick={ this.logoutClick }>Logout</Button>
-              : <Button type="submit" className="btn modal-btn auth authorize">Authorize</Button>
+                nonOauthDefinitions.size === authorizedAuth.size
+                  ? <Button className="sui-btn sui-btn--tertiary-lt modal-btn auth" onClick={ this.logoutClick } ><span>Logout</span></Button>
```
Comment (shockey) [inline src/core/components/auth/auths.jsx]: **Change request**: rename the `auth` class so that it makes sense w/r/t BEM in the context of this button (again, no idea what it's related to or trying to convey)

(PS: I know you didn't introduce the `auth` class, or `authorize` for that matter, but if we're going to stick with them in our new button implementation we should address them)

File: src/core/components/auth/auths.jsx [hunk 5]
```diff
@@ -80,12 +80,13 @@ export default class Auths extends React.Component {
...
+                  : <Button type="submit" className="sui-btn sui-btn--primary modal-btn auth authorize"><span>Authorize</span></Button>
               }
-              <Button className="btn modal-btn auth btn-done" onClick={ this.close }>Close</Button>
+              <Button className="sui-btn sui-btn--secondary modal-btn auth btn-done" onClick={ this.close }><span>Close</span></Button>
```
Comment (shockey) [inline src/core/components/auth/auths.jsx]: **Question**: should `btn-done` be prefixed with `sui` and/or should `done` be a modifier of `btn`?

File: src/core/plugins/oas3/components/request-body-editor.jsx [hunk 6]
```diff
@@ -116,14 +116,22 @@ export default class RequestBodyEditor extends PureComponent {
...
+                    className={`sui-btn sui-btn--secondary body-param__example-edit ${isEditBox ? "cancel" : "edit"}`}
+                    onClick={this.toggleIsEditBox}
+                  >
+                    <span>{isEditBox ? "Cancel" : "Edit"}</span>
+                  </Button>
             }
             { userDidModify &&
-              <Button className="btn ml3" onClick={() => { this.resetValueToSample(mediaType) }}>Reset</Button>
+              <Button
+                className="sui-btn sui-btn--secondary ml3"
```
Comment (shockey) [inline src/core/plugins/oas3/components/request-body-editor.jsx]: **Change request**: drop `ml3` (it's a [Tachyons class](http://tachyons.io/docs/layout/spacing/) that I really shouldn't have kept around, cc #4460), set a `margin-left` for the button as needed instead

File: src/core/components/auth/auths.jsx (hunk 5 above)
Comment (Rody-Kirwan) [reply src/core/components/auth/auths.jsx]: Possibly... I'd say more likely that it is no longer needed. Will check an tidy up as necessary :)

File: src/core/components/auth/auths.jsx (hunk 4 above)
Comment (Rody-Kirwan) [reply src/core/components/auth/auths.jsx]: > (PS: I know you didn't introduce the `auth` class, or `authorize` for that matter, but if we're going to stick with them in our new button implementation we should address them)

Absolutely. Will update :)

Comment (Rody-Kirwan) [reply src/core/components/auth/auths.jsx]: @shokey Seems neither classes are required here by CSS but both are used by cypress. 
```
.get("button.btn.modal-btn.auth.authorize.button")
```
Maybe element IDs would be more appropriate here?

File: src/core/components/operation.jsx [hunk 7]
```diff
@@ -183,7 +184,11 @@ export default class Operation extends PureComponent {
                   </div> : null
               }
 
-            <div className={(!tryItOutEnabled || !response || !allowTryItOut) ? "execute-wrapper" : "btn-group"}>
+            <div
+              className={cx("execute-wrapper", {
+                "sui-btn-group": (!tryItOutEnabled || !response || !allowTryItOut)
```
Comment (shockey) [inline src/core/components/operation.jsx]: **Question**: this is a functionality change, it wasn't formerly possible to have `execute-wrapper` and `sui-btn-group` applied at the same time. Intentional?

Comment (Rody-Kirwan) [reply src/core/components/operation.jsx]: Yes - it was intentional, but maybe not the best change. I wanted to apply the styles used on `.execute wrapper` to the `btn-group` to avoid adding a padding default to a reusable css class. Maybe it is better to create a specific wrapper for the `sui-btn-group`?

File: test/e2e-cypress/tests/features/oauth2-flows/password.js [hunk 8]
```diff
@@ -30,7 +30,7 @@ describe("OAuth2 Password flow", function() {
       .clear()
       .type("secret")
 
-      .get("div.modal-ux-content > div:nth-child(1) > div > div:nth-child(2) > div > div.auth-btn-wrapper > button.sui-btn.modal-btn.authorize")
+      .get("div.modal-ux-content > div:nth-child(1) > div > div:nth-child(2) > div > div.auth-btn-wrapper > button.sui-btn--primary")
```
Comment (Rody-Kirwan) [inline test/e2e-cypress/tests/features/oauth2-flows/password.js]: Yup. Didn't think of that :D 
I take it you'd like to avoid the use of IDs as much as possible?

Comment (shockey) [reply test/e2e-cypress/tests/features/oauth2-flows/password.js]: yeah, a couple of reasons:
* you can only have one element with an ID in your DOM... adopting non-dynamic IDs could cause problems for folks that have multiple Swagger UI instances on their page at once
* having IDs around encourages folks to extend Swagger UI by using native listeners and such that target the native element by ID, rather than using our plugin system

**Comment by shockey** [comment]
thanks @Rody-Kirwan!
